File: composeApp/src/androidMain/kotlin/ule/jescuj00/fridgey/ui/screens/ajustes/AjustesViewModel.kt
package ule.jescuj00.fridgey.ui.screens.ajustes

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import ule.jescuj00.fridgey.data.repository.PreferenciasRepository
import ule.jescuj00.fridgey.domain.model.ModoAnadirProducto
import ule.jescuj00.fridgey.domain.notification.NotificacionCaducidadScheduler

data class AjustesUiState(
    // Refleja la INTENCIÓN del usuario (la preferencia), no el permiso del SO.
    val avisosCaducidad: Boolean = true,
    // Modo con el que el botón "+" de una nevera abre el alta de producto.
    val modoAnadir: ModoAnadirProducto = ModoAnadirProducto.DEFAULT,
)

class AjustesViewModel(
    private val preferenciasRepository: PreferenciasRepository,
    private val scheduler: NotificacionCaducidadScheduler,
) : ViewModel() {

    private val _uiState = MutableStateFlow(AjustesUiState())
    val uiState: StateFlow<AjustesUiState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch {
            val avisos = preferenciasRepository.avisosCaducidadActivados()
            val modo = preferenciasRepository.modoAnadirProducto()
            _uiState.update { it.copy(avisosCaducidad = avisos, modoAnadir = modo) }
        }
    }

    /**
     * Persiste el modo de añadido por defecto del botón "+". Preferencia LOCAL,
     * sin efectos secundarios (no programa nada): solo guarda la elección.
     */
    fun onModoAnadirSeleccionado(modo: ModoAnadirProducto) {
        _uiState.update { it.copy(modoAnadir = modo) }
        viewModelScope.launch {
            preferenciasRepository.setModoAnadirProducto(modo)
        }
    }

    /**
     * Persiste la intención del usuario y (des)programa el motor:
     *  - ON  -> guarda true, programa el periódico (KEEP) + un barrido inmediato.
     *  - OFF -> guarda false, cancela el periódico.
     *
     * Activar pone la preferencia a true AUNQUE falte el permiso del SO: el
     * Worker ya lo gestiona (mostrar() devuelve false, no marca) y los avisos
     * arrancan solos cuando se conceda. La PETICIÓN de permiso la dispara el
     * Composable (HITO 4B), nunca el ViewModel.
     */
    fun onToggle(activado: Boolean) {
        _uiState.update { it.copy(avisosCaducidad = activado) }
        viewModelScope.launch {
            preferenciasRepository.setAvisosCaducidadActivados(activado)
            if (activado) {
                scheduler.programarComprobacionDiaria()
                scheduler.comprobarAhora()
            } else {
                scheduler.cancelar()
            }
        }
    }

    /**
     * Resultado de la petición de permiso (la LANZA el Composable, no el VM).
     * Marca que ya se solicitó y, si se concedió, dispara un barrido inmediato
     * para no esperar al ciclo diario.
     */
    fun onPermisoResultado(concedido: Boolean) {
        viewModelScope.launch {
            preferenciasRepository.setPermisoNotifSolicitado(true)
            if (concedido) scheduler.comprobarAhora()
        }
    }
}

File: iosApp/iosApp/Screens/Ajustes/AjustesViewModel.swift
import Foundation
import UserNotifications
import Shared

/// Estado y lógica de la pantalla de Ajustes (Fase 1b iOS). El toggle refleja la
/// INTENCIÓN del usuario (preferencia); el permiso del SO es una compuerta aparte.
@MainActor
final class AjustesViewModel: ObservableObject {

    @Published var avisosCaducidad: Bool = true   // default ON
    @Published var permisoDenegado: Bool = false  // authorizationStatus == .denied
    // Modo con el que el botón "+" de una nevera abre el alta de producto.
    @Published var modoAnadir: ModoAnadirProducto = .manual

    private let preferencias = KoinIosKt.getPreferenciasRepository()
    private let controller = AvisosCaducidadController.shared
    private let center = UNUserNotificationCenter.current()

    func cargar() async {
        avisosCaducidad = (try? await preferencias.avisosCaducidadActivados())?.boolValue ?? true
        modoAnadir = (try? await preferencias.modoAnadirProducto()) ?? .manual
        await refrescarPermiso()
    }

    /// Persiste el modo de añadido por defecto del botón "+". Preferencia LOCAL,
    /// sin efectos secundarios (no programa nada): solo guarda la elección.
    func onModoAnadirSeleccionado(_ modo: ModoAnadirProducto) async {
        modoAnadir = modo
        try? await preferencias.setModoAnadirProducto(modo: modo)
    }

    /// Refresca el flag de denegación (para el banner). Se llama al cargar y al
    /// volver de los ajustes del sistema (scenePhase .active).
    func refrescarPermiso() async {
        permisoDenegado = await center.notificationSettings().authorizationStatus == .denied
    }

    /// Persiste la intención y (des)programa. Al activar sin permiso decidido,
    /// pide el permiso (en iOS la petición es directa, no necesita launcher como
    /// Android). El Worker/controlador ya gestiona el caso "sin permiso".
    func onToggle(_ activado: Bool) async {
        avisosCaducidad = activado
        try? await preferencias.setAvisosCaducidadActivados(activado: activado)
        if activado {
            if await center.notificationSettings().authorizationStatus == .notDetermined {
                _ = await controller.solicitarPermiso()
            }
            await refrescarPermiso()
        }
        await controller.reconciliar()
    }
}
